src/lib.rs

```diff
@@ -96,28 +96,43 @@ pub use once_cell::sync::Lazy;
 #[macro_export]
 macro_rules! flate {
     ($(#[$meta:meta])*
-        $(pub $(($($vis:tt)+))?)? static $name:ident: [u8] from $path:literal $(with $algo:ident)?) => {
+        $(pub $(($($vis:tt)+))?)? static $name:ident: [u8] from $path:literal $(with $algo:ident)? $(if $($threshold:tt)+)?) => {
         // HACK: workaround to make cargo auto rebuild on modification of source file
         const _: &'static [u8] = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/", $path));
 
         $(#[$meta])*
         $(pub $(($($vis)+))?)? static $name: $crate::Lazy<::std::vec::Vec<u8>> = $crate::Lazy::new(|| {
-            $crate::decode($crate::codegen::deflate_file!($path), None)
+            // Evaluate the condition at compile time to avoid unnecessary runtime checks
+            if $crate::codegen::deflate_if!($path $($algo)? $($($threshold)+)?) {
+                let algo = match stringify!($($algo)?){
+                    "deflate" => $crate::CompressionMethod::Deflate,
+                    "zstd" => $crate::CompressionMethod::Zstd,
+                    _ => $crate::CompressionMethod::default(),
+                };
+                $crate::decode($crate::codegen::deflate_file!($path $($algo)?), Some($crate::CompressionMethodTy(algo)))
+            } else {
+                include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/", $path)).to_vec()
+            }
         });
     };
     ($(#[$meta:meta])*
-        $(pub $(($($vis:tt)+))?)? static $name:ident: str from $path:literal $(with $algo:ident)?) => {
+        $(pub $(($($vis:tt)+))?)? static $name:ident: str from $path:literal $(with $algo:ident)? $(if $($threshold:tt)+)?) => {
         // HACK: workaround to make cargo auto rebuild on modification of source file
         const _: &'static str = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/", $path));
 
         $(#[$meta])*
         $(pub $(($($vis)+))?)? static $name: $crate::Lazy<::std::string::String> = $crate::Lazy::new(|| {
-            let algo = match stringify!($($algo)?){
-                "deflate" => $crate::CompressionMethod::Deflate,
-                "zstd" => $crate::CompressionMethod::Zstd,
-                _ => $crate::CompressionMethod::default(),
-            };
-            $crate::decode_string($crate::codegen::deflate_utf8_file!($path $($algo)?), Some($crate::CompressionMethodTy(algo)))
+            // Evaluate the condition at compile time to avoid unnecessary runtime checks
+            if $crate::codegen::deflate_if!($path $($algo)? $($($threshold)+)?) {
```

Comment:
Could we actually use conditional compilation here? While the compiler should optimize away the constant false branch, it seems weird that we are including the file into the program twice just so that we could remove it later.

Reply:
E.g. can we make `deflate_if!` emit `all()` for true and `any()` for false instead?

Reply:
Nope. Unfortunately, the [conditional compilation](https://doc.rust-lang.org/reference/conditional-compilation.html) does not support such statements.

This (relying on compiler optimization with constant boolean statement) is the only way to achieve this without huge code refactor as described in the PR description.

Although it is still possible by either one of:
- making `flate!` itself a proc-macro.
- make a proc-macro in codegen crate that generates
  - if the `deflate_if!` returns **true**:
    - `$(pub $(($($vis)+))?)? static $name: $crate::Lazy<::std::vec::Vec<u8>> = $crate::Lazy::new(...)`; or
  - if the `deflate_if!` returns **false**:
    - `$(pub $(($($vis)+))?)? static $name: &[u8] = include_str!(...)`.

Either ways are much more huge refactor/codebase change than just let the compiler optimize out.

Reply:
> it seems weird that we are including the file into the program twice just so that we could remove it later.

Please note that `deflate_if!` does not actually deflates the file. It is used for deciding whether or not the compression ratio meets the given threshold criteria. So the only a single file will be embedded in the binary, not the twice.